x/ecocredit/client/testsuite/tx.go

```diff
@@ -1426,6 +1426,138 @@ func (s *IntegrationTestSuite) TestTxBuyDirect() {
+		s.Run(tc.name, func() {
+			if len(tc.errMsg) != 0 {
+				_, err := cli.ExecTestCLICmd(clientCtx, cmd, tc.args)
+				s.Require().ErrorContains(err, tc.errMsg)
+			} else {
+				sellerAccBefore := s.getAccountInfo(clientCtx, val0.Address, askCoin.Denom, batchDenom)
+				buyerAccBefore := s.getAccountInfo(clientCtx, buyerAcc, askCoin.Denom, batchDenom)
+
+				out, err := cli.ExecTestCLICmd(clientCtx, cmd, tc.args)
+				s.Require().NoError(err)
+
+				sellerAccAfter := s.getAccountInfo(clientCtx, val0.Address, askCoin.Denom, batchDenom)
+				buyerAccAfter := s.getAccountInfo(clientCtx, buyerAcc, askCoin.Denom, batchDenom)
+
+				var res sdk.TxResponse
+				s.Require().NoError(clientCtx.Codec.UnmarshalJSON(out.Bytes(), &res))
+				s.Require().Equal(uint32(0), res.Code)
+				s.assertMarketBalanceBatchUpdated(sellerAccBefore, sellerAccAfter, buyerAccBefore, buyerAccAfter, buyOrders)
+			}
+		})
+	}
+}
+
+// assertMarketBalanceBatchUpdated asserts that all accounts involved in a marketplace transaction are updated properly.
+// it assumes that both seller/buyer accounts used the same denom for amount sold/bought.
+func (s *IntegrationTestSuite) assertMarketBalanceBatchUpdated(sb, sa, bb, ba accountInfo, orders []*marketplace.MsgBuyDirect_Order) {
+	tradSold, retSold := math.NewDecFromInt64(0), math.NewDecFromInt64(0)
+	cost := sdk.NewInt64Coin(sdk.DefaultBondDenom, 0)
+	for _, order := range orders {
+		qty, err := math.NewDecFromString(order.Quantity)
+		s.Require().NoError(err)
+		if order.DisableAutoRetire {
+			tradSold, err = tradSold.Add(qty)
+			s.Require().NoError(err)
+		} else {
+			retSold, err = retSold.Add(qty)
+			s.Require().NoError(err)
+		}
+		costDec, err := math.NewDecFromString(order.BidPrice.Amount.String())
+		s.Require().NoError(err)
+		totalCostDec, err := qty.Mul(costDec)
+		s.Require().NoError(err)
+		c := sdk.NewCoin(order.BidPrice.Denom, totalCostDec.SdkIntTrim())
+		cost = cost.Add(c)
+	}
+
+	totalSold, err := tradSold.Add(retSold)
+	s.Require().NoError(err)
+
+	// check sellers coins
+	expectedSellerGain := sb.coinBal.Add(cost)
+	s.Require().True(expectedSellerGain.Equal(sa.coinBal))
+
+	// check buyers coins
+	// we use LT in the buyer case, as some coins go towards fees, so their balance will be LOWER than before - total cost.
+	expectedBuyerCost := bb.coinBal.Sub(cost)
+	s.Require().True(ba.coinBal.IsLT(expectedBuyerCost))
+
+	// check sellers credits
+	expectedEscrowed, err := sb.escrowed.Sub(totalSold)
+	s.Require().NoError(err)
+	s.Require().True(expectedEscrowed.Equal(sa.escrowed))
+	s.Require().True(sb.tradable.Equal(sa.tradable))
+	s.Require().True(sb.retired.Equal(sa.retired))
+
+	expectedRetired, err := bb.retired.Add(retSold)
+	s.Require().NoError(err)
+	expectedTradable, err := bb.tradable.Add(tradSold)
+	s.Require().NoError(err)
+
+	s.Require().True(bb.escrowed.Equal(bb.escrowed))
+	s.Require().True(ba.tradable.Equal(expectedTradable))
+	s.Require().True(ba.retired.Equal(expectedRetired))
```

Comment:
Any reason we're not using `s.Require().Equal` here? See https://github.com/regen-network/regen-ledger/pull/1079#discussion_r875075437

Reply:
[answered](https://github.com/regen-network/regen-ledger/pull/1079#discussion_r875357729)

Reply:
Would it make more sense to use `String()` which will return the actual and expected results on failing? `True` only returns `expected true` or `expected false`.

Reply:
updated in https://github.com/regen-network/regen-ledger/pull/1103/commits/0d38b689c5c5e7a27ae1a9c962377061041a671f